transformer_engine/pytorch/module/layernorm_linear.py

```diff
@@ -132,7 +132,7 @@ def forward(
                 # which will be later copied to a FP8 UB
                 ln_out = torch.empty_like(inputmat, memory_format=torch.contiguous_format)
             else:
-                ln_out = ub_obj_lnout.get_ubuf_output(0)
+                ln_out = ub_obj_lnout.get_ubuf_output(tex.NVTE_Comm_Overlap_Type.RS)
```

Comment:
I don't fully understand this logic, but it seems like the layernorm output shouldn't involve reduce-scatters? That said, I see in the old impl that `ln_out` and `ln_out_total` are different buffers from `get_ubuf_output`. If my suggestion is incorrect, then it would be better to revert to using `int`s instead of enums.
```suggestion
                ln_out = ub_obj_lnout.get_ubuf_output(tex.NVTE_Comm_Overlap_Type.AG)
```

Reply:
The logic before Alp's refactoring is already implicit. I also got confused before.
The point is: Both `ln_out` and `ln_out_total` shares the same underlying buffer of a tensor, but may refer to different parts.
e.g. 
- `ln_out = get_ubuf_output(tex.NVTE_Comm_Overlap_Type.RS)` divides the buffer into chunks and `ln_out` refer to each chunk head depending different tp-rank it is
- `ln_out_total = get_ubuf_output(tex.NVTE_Comm_Overlap_Type.AG)` refer to the head of the whole buffer

so in python layer, the original code before refactoring is confusing because:
```python
ln_out, mu, rsigma = _apply_normalization( # write *something* to each chunk of the buffer
    inputmat,
    ln_out,
    ....
)
ln_out_total = ub_obj_lnout.get_ubuf_output(tex.NVTE_Comm_Overlap_Type.AG) # get the whole buffer, including *something* 
if not return_layernorm_output:
  ln_out = torch.empty_like(ln_out)
  # ln_out is dereferenced in python, but the underlying buffer has already partially changed, and still can be touched by ln_out_total
  # here comes the confusing part ^^^
```

Reply:
The logic here is related to what's in the communication buffer on each rank *after* the comm+GEMM overlap finishes execution.

After the AG overlap executes, the communication buffer contains the all-gathered global data on every rank. Therefore, `get_ubuf_output(AG)` returns a PyTorch tensor that spans the entire communication buffer.

After the RS overlap executes, the chunk of the communication buffer belonging to each rank contains that rank's local portion of the distributed GEMM output. Therefore, `get_ubuf_output(RS)` returns a PyTorch tensor that only spans this local chunk of the buffer.

I personally found the old integer argument here to be much more opaque and confusing. Less subjectively, though, an integer type for this argument hides the fact that only 0 and 1 are valid choices. A boolean flag would be acceptable, but I still think the AG/RS enums provide the most descriptive interface.

Reply:
I see, so logically this should be an enum for the buffer type, not the overlap type. My preference is enum > int > misleading enum, so making a new enum type with something like `SHARDED`/`GATHERED` would be better than `AG`/`RS`.

Reply:
I'm on board with a new enum for this. What do you think about `NVTE_Comm_Overlap_Output` with `GLOBAL` and `LOCAL` values? `GATHERED` is a bit too close to `ALL_GATHER`, and there are some cases where we want to get the global buffer even when it has not yet been gathered.